{% extends "layout.html" %}

{% block title %}
About
{% endblock title %}

{% block body %}
<div class='container'>
    <div class="col-md-12">
        <h1>About the Swedish ASV portal</h1>
        <p>
            The portal of Swedish Amplicon Sequence Variants (ASVs)
            is an interface to sequence-based observations in
            the <a href="{{ sbdi_start_page }}">
            Swedish Biodiversity Data Infrastructure (SBDI)</a>.
            In addition to a database of amplicon sequence variants (ASVs) and a web interface,
            we provide a
            <a href="https://biodiversitydata-se.github.io/mol-data/ena-metabar.html"
               rel="noreferrer noopener" target="_blank">step-by-step guide<span class='screen-reader-text'>(Link opens in a new tab)</span></a>
            to submitting raw sequence data to ENA,
            and a
            <a href="https://nf-co.re/ampliseq" target="_blank"
               rel="noreferrer noopener">pipeline
               <span class='screen-reader-text'>(Link opens in a new tab)</span>
           </a>
            (developed in the nf-core collaboration)
            for denoising and taxonomic annotation of metabarcoding data.
        </p>
        <p>
            <figure>
                <img src="{{url_for('static', filename='img/metabarcoding-sbdi.png')}}"
                     class="img-fluid" title="Flow of metabarcoding data in SBDI" />
                <figcaption><br><b>Flow of metabarcoding data in SBDI</b>: A sequencing service provider, such as NGI, delivers sequencing data
                    to the user. The user submits raw sequencing data and contextual data to ENA
                    (see <a href="https://biodiversitydata-se.github.io/mol-data/ena-metabar.html"
                    rel="noreferrer noopener">our guide<span class='screen-reader-text'>(Link opens in a new tab)</span></a>), denoises the data using e.g.
                    <a href="https://nf-co.re/ampliseq" rel="noreferrer noopener" target="_blank">nf-core/ampliseq
                        <span class='screen-reader-text'>(Link opens in a new tab)</span>
                    </a>, and submits the denoised data (ASVs and their counts in different samples, and contextual data)
                    to the ASV portal. Metabarcoding data in the ASV database/Bioatlas
                    can be searched and downloaded using the portal or using the
                    <a href="https://github.com/biodiversitydata-se/SBDI4R"
                    rel="noreferrer noopener">SBDI4R package</a> in R.
               </figcaption>
            </figure>
        <p>
            <h2>Why publish data in SBDI?</h2>
            <p>By publishing your data in the Swedish ASV database and Bioatlas
            your data will be easily reusable by others and can be seamlessly analysed
            in conjunction with other datasets in the Bioatlas. The ASVs will be
            regularly re-annoted as new reference database versions are released;
            thus your data will be updated with the latest taxonomic information.
            The ASV database also provides a complement to sequence repositories
            such as ENA or NCBI that typically only store raw sequencing data.
            Data in the Bioatlas is further regularly integrated in GBIF and therefor
            your metabarcoding-based biodiversity data will contribute to global
            knowledge about biodiversity.</p>
        </p>
        <p>
            <h2>Available data</h2>
            <p>Currently, we support inclusion of ASVs for the 16S rRNA gene for prokaryotes,
            and the internal transcribed spacer (ITS) for fungi. Later in 2021 we will
            support the cytochrome c oxidase I (COI) for animals and the 18S rRNA gene
            for eukaryotes. Other markers can be included on request.
            Read more about ASV taxonomy in SBDI
            <a href="https://docs.biodiversitydata.se/analyse-data/molecular-tools/">here</a>.</p>
        </p>
        <p>
            <caption><b>Marker genes and taxa currently represented in the ASV database</b></caption>

            <br>
            <table id='stats_table' class='dataframe table table-striped'>
                <thead>
                    <tr>
                        <th>Marker</th>
                        <th>Domains/Kingdoms</th>
                        <th># Datasets</th>
                        <th># Phyla</th>
                        <th># Classes</th>
                        <th># Orders</th>
                        <th># Families</th>
                        <th># Genera</th>
                        <th># Species</th>
                        <th># ASV:s</th>
                    </tr>
                </thead>
                    {% if rows %}
                    {% for row in rows %}
                    <tr>
                        <td>{{ row.gene }}</a></td>
                        <td>{{ row.kingdoms }}</a></td>
                        <td align="center">{{ row.datasets }}</a></td>
                        <td align="center">{{ row.phyla }}</a></td>
                        <td align="center">{{ row.classes }}</a></td>
                        <td align="center">{{ row.orders }}</a></td>
                        <td align="center">{{ row.families }}</a></td>
                        <td align="center">{{ row.genera }}</a></td>
                        <td align="center">{{ row.species }}</a></td>
                        <td align="center">{{ row.asvs }}</a></td>
                    </tr>
                    {% endfor %}
                    {% endif %}
            </table>
        </p>
        <br>
        <p>
            We are currently working on improving the accessibility of this website.
            <a href="{{ sbdi_contact_page }}">Please contact SBDI support</a> with any questions or suggestions!
        </p>

</div>
{% endblock body %}
