{% extends 'layout.html' %}
{% block body %}
<div class='container'>
    <div>
        <h1>BLAST for ASVs and Bioatlas records</h1>
        <p>Use the Basic Local Alignment Search Tool (BLAST), to find Amplicon
           Sequence Variants (ASVs), and associated Bioatlas occurrence records,
           <br>matching your sequences and alignment criteria.
           Read more about how to search by sequence <a href="{{ sbdi_seq_search_page }}">here</a>.
       </p>
    <hr>
    </div>
    <form action='' method='post' id='sform' name='sform' class='search_function'>
        {{ sform.csrf_token }}
        <div class='row'>
            <div class='col-md-12'>
                {% if sform.sequence.errors %}
                    <div class='form-group has-error'>
                {% else %}
                    <div class='form-group'>
                {% endif %}
                        <p>
                            <b>Query sequence(s)</b>
                            <br>Nucleotide sequence(s), in fasta format, to compare against ASVs (subject sequences) in reference database
                            <span style='float: right' id='sequence_count'>2108/50000 characters</span>
                        </p>
                            {{ sform.sequence(id='sequence_textarea', rows=10, cols=80, maxlength='50000', class='form-control') }}
                            <span id='helpBlock2' class='help-block'>
                                {% for error in sform.sequence.errors %}
                                {{ error }}
                                {% endfor %}
                            </span>
                    </div>
            </div>
        </div>
        <br>
        <div class='row'>
            <div class='col-md-3'>
                <p>
                    <b>Min. identity (Id %)</b>
                        <br>Share of exact matches in alignment
                </p>
                {% if sform.min_identity.errors %}
                    <div class='form-group has-error'>
                {% else %}
                    <div class='form-group'>
                {% endif %}
                        {{ sform.min_identity(id='min_identity_input', class='form-control', type='number') }}
                        <span class='help-block'>
                            {% for error in sform.min_identity.errors %}
                            {{ error }}
                            {% endfor %}
                        </span>
                    </div>
            </div>
            <div class='col-md-3'>
                <p>
                    <b>Min. query coverage (Cov %)</b>
                    <br>Share of aligned query bases
                </p>
                {% if sform.min_qry_cover.errors %}
                    <div class='form-group has-error'>
                {% else %}
                    <div class='form-group'>
                {% endif %}
                        {{ sform.min_qry_cover(id='min_qry_cover_input', class='form-control', type='number') }}
                        <span class='help-block'>
                            {% for error in sform.min_qry_cover.errors %}
                            {{ error }}
                            {% endfor %}
                        </span>
                    </div>
            </div>
        </div>
        <div class='row'>
            <div class='col-md-12'>
                This corresponds to running a (command-line) nucleotide BLAST (<i>BLASTn</i>) search filtered
                on <i>-perc_identity</i> and <i>-qcov_hsp_perc</i>.
                Note that query cover is calculated per <i>High Scoring Pair (HSP)</i>,
                i.e. per ungapped local pairwise alignment between query and subject,
                and that we only report the best HSP for each query-subject hit.
                Result fields correspond to <i>qacc</i>, <i>sacc</i>, <i>pident</i>, <i>qcovhsp</i> and <i>evalue</i>
                in tabular BLASTn output format (<i>â€“outfmt=6</i>),
                and rows are sorted on Query and E-value by default.
                The E-value specifies the number of hits of similar (or higher) score
                that we expect to see by chance, i.e. smaller values indicate 'better' hits.
            </div>
        </div>

        <br>
        <hr>
        <div class='row'>
            <div class='col-md-3'>
                {{ sform.blast_for_seq(class='btn btn-primary', id='blast_for_seq') }}
                &nbsp;&nbsp;
                <input type='button' class='btn btn-primary' id='clear_all' value='Clear'/>
            </div>
            <div class='col-md-9'>
                You may have to <span class='scroll-note'>scroll down to see results</span>.
                When they have finished loading, tick left column to (de)select a row,
                and click bottom button to show occurrence records of selected ASV:s in Bioatlas.
                You can also download results as Excel/CSV, and link these to Bioatlas records via <i>taxonID</i> column.
            </div>
        </div>
    </form>
</div>
<br>
<br>

{% if rform %}
<div class='container' id='result'>
    <form action={{ rform.batch_url }} method='post' id='rform' name='rform'>
        {{ rform.csrf_token }}
        <div class='row'>
            <div class='col-md-12'>
                <table id='blast_result_table' class='dataframe table table-striped'>
                    <thead>
                        <tr>
                            <th style="text-align: center"><input type="checkbox" id="select_all" /></th>
                            <th>taxonID (Bioatlas link)</th>
                            <th>Query</th>
                            <th>Subject</th>
                            <th>Id (%)</th>
                            <th>Cov (%)</th>
                            <th>E-value</th>
                            <th>Subject sequence</th>
                        </tr>
                    </thead>
                </table>
            </div>
        </div>
        {% include "sbdi_post.html" %}
    </form>
</div>
{% endif %}

{% endblock body %}
